File: pilotdeck-vdi/freecad/vdi_cad_core/dexpi_export.py
"""DEXPI Proteus XML subset export from PlantModel v1."""

import json
import os
import re
from datetime import datetime, timezone
from typing import Any, Dict, List, Optional
from xml.etree.ElementTree import Element, SubElement, tostring

DEXPI_NS = "http://www.dexpi.org/PlantModel/1.4"
RDL = "http://data.posccaesar.org/rdl"

EQUIPMENT_CLASS = {
    "pump": ("CentrifugalPump", f"{RDL}/RDS416937"),
    "vessel": ("Vessel", f"{RDL}/RDS414674"),
    "tank": ("Tank", f"{RDL}/RDS414674"),
    "reactor": ("Vessel", f"{RDL}/RDS414674"),
    "column": ("Column", f"{RDL}/RDS415842"),
    "heat_exchanger": ("ShellAndTubeHeatExchanger", f"{RDL}/RDS416321"),
    "compressor": ("Compressor", f"{RDL}/RDS415842"),
    "generic": ("Equipment", f"{RDL}/RDS414674"),
}

VALVE_CLASS = {
    "gate": ("GateValve", f"{RDL}/RDS416842"),
    "ball": ("BallValve", f"{RDL}/RDS416842"),
    "check": ("CheckValve", f"{RDL}/RDS416842"),
}


def _mapping_path() -> str:
    return os.path.normpath(
        os.path.join(os.path.dirname(__file__), "..", "..", "mcp", "vdi-cad", "schemas", "cfihos-vdi-mapping.json")
    )


def load_cfihos_mapping() -> Dict[str, Any]:
    with open(_mapping_path(), encoding="utf-8") as f:
        return json.load(f)


def _dexpi_id(object_id: str, prefix: str = "obj") -> str:
    clean = re.sub(r"[^a-zA-Z0-9]", "", object_id or "")
    return f"{prefix}_{clean[:32] or 'unknown'}"


def _nozzle_id(equipment_id: str, role: str) -> str:
    return f"{_dexpi_id(equipment_id, 'eq')}_nozzle_{role}"


def _add_generic_attributes(parent: Element, attrs: Dict[str, Any], mapping: Dict[str, Any]) -> None:
    ga_list: List[Dict[str, str]] = []
    for key, val in (attrs or {}).items():
        if val is None or val == "":
            continue
        meta = mapping.get(key, {})
        entry: Dict[str, str] = {"Name": str(meta.get("dexpi") or key), "Value": str(val)}
        unit = meta.get("unit")
        if unit:
            entry["Format"] = unit
        ga_list.append(entry)
    if not ga_list:
        return
    ga_wrap = SubElement(parent, "GenericAttributes", Number=str(len(ga_list)))
    for entry in ga_list:
        SubElement(ga_wrap, "GenericAttribute", **entry)


def plant_model_to_dexpi_xml(model: Dict[str, Any], discipline: str = "PID") -> str:
    mapping = load_cfihos_mapping()
    now = datetime.now(timezone.utc)
    project_id = model.get("project_id") or "VDI-PROJECT"
    revision = model.get("revision") or "A"
    objects = model.get("objects") or []

    equipment = [o for o in objects if o.get("class") == "Equipment"]
    pipes = [o for o in objects if o.get("class") == "PipeRun"]
    instruments = [o for o in objects if o.get("class") == "Instrument"]
    valves = [o for o in objects if o.get("class") == "Valve"]
    psvs = [o for o in objects if o.get("class") == "SafetyValve"]

    root = Element("PlantModel")
    root.set("xmlns", DEXPI_NS)

    pi = SubElement(root, "PlantInformation")
    pi.set("OriginatingSystem", "PilotDeck-VDI")
    pi.set("OriginatingSystemVersion", "1.0")
    pi.set("Date", now.strftime("%Y-%m-%d"))
    pi.set("Time", now.strftime("%H:%M:%S"))
    pi.set("Is3D", "no")
    pi.set("Discipline", discipline)
    pi.set("SchemaVersion", "4.2.0")
    SubElement(pi, "UnitsOfMeasure", Distance="Millimetre", Pressure="MPa", Temperature="DegreeCelsius")
    ga_pi = SubElement(pi, "GenericAttributes", Number="2")
    SubElement(ga_pi, "GenericAttribute", Name="ProjectId", Value=project_id)
    SubElement(ga_pi, "GenericAttribute", Name="Revision", Value=revision)

    eq_map = mapping.get("mappings", {}).get("Equipment", {})
    for eq in equipment:
        attrs = eq.get("attributes") or {}
        et = str(attrs.get("equipment_type", "generic")).lower()
        cls, uri = EQUIPMENT_CLASS.get(et, EQUIPMENT_CLASS["generic"])
        el = SubElement(root, "Equipment")
        el.set("ID", _dexpi_id(eq["object_id"], "eq"))
        el.set("TagName", eq.get("tag", ""))
        el.set("ComponentClass", cls)
        el.set("ComponentClassURI", uri)
        _add_generic_attributes(el, attrs, eq_map)
        SubElement(
            el,
            "Nozzle",
            ID=_nozzle_id(eq["object_id"], "in"),
            TagName="N1",
            ComponentClass="Nozzle",
            ComponentClassURI=f"{RDL}/RDS415214",
        )
        SubElement(
            el,
            "Nozzle",
            ID=_nozzle_id(eq["object_id"], "out"),
            TagName="N2",
            ComponentClass="Nozzle",
            ComponentClassURI=f"{RDL}/RDS415214",
        )

    if pipes:
        pns = SubElement(root, "PipingNetworkSystem")
        pns.set("ID", "pns_main")
        pns.set("TagName", f"{project_id}-Piping")
        pns.set("ComponentClass", "PipingNetworkSystem")
        pns.set("ComponentClassURI", f"{RDL}/RDS270359")
        pipe_map = mapping.get("mappings", {}).get("PipeRun", {})
        for pipe in pipes:
            attrs = pipe.get("attributes") or {}
            seg = SubElement(pns, "PipingNetworkSegment")
            seg_id = _dexpi_id(pipe["object_id"], "seg")
            seg.set("ID", seg_id)
            seg.set("TagName", pipe.get("tag", ""))
            seg.set("ComponentClass", "PipingNetworkSegment")
            seg.set("ComponentClassURI", f"{RDL}/RDS267704")
            _add_generic_attributes(seg, attrs, pipe_map)
            from_tag = attrs.get("from_tag", "")
            to_tag = attrs.get("to_tag", "")
            from_eq = next((e for e in equipment if e.get("tag") == from_tag), None)
            to_eq = next((e for e in equipment if e.get("tag") == to_tag), None)
            if from_eq:
                SubElement(seg, "Connection", FromID=_nozzle_id(from_eq["object_id"], "out"), ToID=f"{seg_id}_start")
            if to_eq:
                SubElement(seg, "Connection", FromID=f"{seg_id}_end", ToID=_nozzle_id(to_eq["object_id"], "in"))

    inst_map = mapping.get("mappings", {}).get("Instrument", {})
    for inst in instruments:
        el = SubElement(root, "InstrumentationLoopFunction")
        el.set("ID", _dexpi_id(inst["object_id"], "inst"))
        el.set("TagName", inst.get("tag", ""))
        el.set("ComponentClass", "ProcessInstrumentationFunction")
        el.set("ComponentClassURI", f"{RDL}/RDS416842")
        _add_generic_attributes(el, inst.get("attributes") or {}, inst_map)

    valve_map = mapping.get("mappings", {}).get("Valve", {})
    for valve in valves:
        vt = str((valve.get("attributes") or {}).get("valve_type", "gate")).lower()
        cls, uri = VALVE_CLASS.get(vt, VALVE_CLASS["gate"])
        el = SubElement(root, "OperatedValve")
        el.set("ID", _dexpi_id(valve["object_id"], "valve"))
        el.set("TagName", valve.get("tag", ""))
        el.set("ComponentClass", cls)
        el.set("ComponentClassURI", uri)
        _add_generic_attributes(el, valve.get("attributes") or {}, valve_map)

    psv_map = mapping.get("mappings", {}).get("SafetyValve", {})
    for psv in psvs:
        el = SubElement(root, "SafetyValveOrFitting")
        el.set("ID", _dexpi_id(psv["object_id"], "psv"))
        el.set("TagName", psv.get("tag", ""))
        el.set("ComponentClass", "SafetyValve")
        el.set("ComponentClassURI", f"{RDL}/RDS416842")
        _add_generic_attributes(el, psv.get("attributes") or {}, psv_map)

    return '<?xml version="1.0" encoding="UTF-8"?>\n' + tostring(root, encoding="unicode")


def dexpi_export_summary(model: Dict[str, Any]) -> Dict[str, Any]:
    objects = model.get("objects") or []
    by_class = {}
    for o in objects:
        c = o.get("class", "")
        by_class[c] = by_class.get(c, 0) + 1
    return {
        "project_id": model.get("project_id"),
        "revision": model.get("revision"),
        "equipment": by_class.get("Equipment", 0),
        "pipe_segments": by_class.get("PipeRun", 0),
        "instruments": by_class.get("Instrument", 0),
        "valves": by_class.get("Valve", 0),
        "safety_valves": by_class.get("SafetyValve", 0),
        "format": "DEXPI-Proteus-subset-1.4",
    }

File: pilotdeck-vdi/mcp/vdi-cad/dexpi-export.mjs
/**
 * DEXPI Proteus XML subset export from PlantModel v1.
 * Covers: Equipment, PipeRun (PipingNetworkSegment), Instrument, Valve, SafetyValve.
 */

import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));

const DEXPI_NS = "http://www.dexpi.org/PlantModel/1.4";
const RDL = "http://data.posccaesar.org/rdl";

const EQUIPMENT_CLASS = {
  pump: { cls: "CentrifugalPump", uri: `${RDL}/RDS416937` },
  vessel: { cls: "Vessel", uri: `${RDL}/RDS414674` },
  tank: { cls: "Tank", uri: `${RDL}/RDS414674` },
  reactor: { cls: "Vessel", uri: `${RDL}/RDS414674` },
  column: { cls: "Column", uri: `${RDL}/RDS415842` },
  heat_exchanger: { cls: "ShellAndTubeHeatExchanger", uri: `${RDL}/RDS416321` },
  compressor: { cls: "Compressor", uri: `${RDL}/RDS415842` },
  generic: { cls: "Equipment", uri: `${RDL}/RDS414674` },
};

const VALVE_CLASS = {
  gate: { cls: "GateValve", uri: `${RDL}/RDS416842` },
  ball: { cls: "BallValve", uri: `${RDL}/RDS416842` },
  check: { cls: "CheckValve", uri: `${RDL}/RDS416842` },
};

function xmlEscape(s) {
  return String(s ?? "")
    .replace(/&/g, "&amp;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;")
    .replace(/"/g, "&quot;");
}

function dexpiId(objectId, prefix = "obj") {
  const clean = String(objectId || "").replace(/[^a-zA-Z0-9]/g, "");
  return `${prefix}_${clean.slice(0, 32) || "unknown"}`;
}

function nozzleId(equipmentId, role) {
  return `${dexpiId(equipmentId, "eq")}_nozzle_${role}`;
}

function genericAttributes(attrs, mapping) {
  const entries = [];
  for (const [key, val] of Object.entries(attrs || {})) {
    if (val == null || val === "") continue;
    const map = mapping?.[key];
    const name = map?.dexpi || key;
    const format = map?.unit || "";
    entries.push(
      `      <GenericAttribute Name="${xmlEscape(name)}" Value="${xmlEscape(val)}"${
        format ? ` Format="${xmlEscape(format)}"` : ""
      }/>`
    );
  }
  if (!entries.length) return "";
  return `    <GenericAttributes Number="${entries.length}">\n${entries.join("\n")}\n    </GenericAttributes>`;
}

let _mappingCache = null;

export function getCfihosMapping() {
  if (!_mappingCache) {
    _mappingCache = JSON.parse(
      fs.readFileSync(path.join(__dirname, "schemas/cfihos-vdi-mapping.json"), "utf8")
    );
  }
  return _mappingCache;
}

/**
 * @param {object} model PlantModel v1
 * @param {{ discipline?: string }} options
 */
export function plantModelToDexpiXml(model, options = {}) {
  const mapping = getCfihosMapping();
  const now = new Date();
  const date = now.toISOString().slice(0, 10);
  const time = now.toTimeString().slice(0, 8);
  const discipline = options.discipline || "PID";
  const revision = model.revision || "A";
  const projectId = model.project_id || "VDI-PROJECT";

  const equipment = (model.objects || []).filter((o) => o.class === "Equipment");
  const pipes = (model.objects || []).filter((o) => o.class === "PipeRun");
  const instruments = (model.objects || []).filter((o) => o.class === "Instrument");
  const valves = (model.objects || []).filter((o) => o.class === "Valve");
  const psvs = (model.objects || []).filter((o) => o.class === "SafetyValve");
  const streams = (model.objects || []).filter((o) => o.class === "Stream");

  const eqMap = mapping.mappings?.Equipment || {};
  const pipeMap = mapping.mappings?.PipeRun || {};
  const instMap = mapping.mappings?.Instrument || {};
  const valveMap = mapping.mappings?.Valve || {};
  const psvMap = mapping.mappings?.SafetyValve || {};

  const parts = [];
  parts.push('<?xml version="1.0" encoding="UTF-8"?>');
  parts.push(`<PlantModel xmlns="${DEXPI_NS}">`);
  parts.push(`  <PlantInformation`);
  parts.push(`    OriginatingSystem="PilotDeck-VDI"`);
  parts.push(`    OriginatingSystemVersion="1.0"`);
  parts.push(`    OriginatingSystemVendor="PilotDeck"`);
  parts.push(`    Date="${date}"`);
  parts.push(`    Time="${time}"`);
  parts.push(`    Is3D="no"`);
  parts.push(`    Discipline="${xmlEscape(discipline)}"`);
  parts.push(`    SchemaVersion="4.2.0">`);
  parts.push(`    <UnitsOfMeasure Distance="Millimetre" Pressure="MPa" Temperature="DegreeCelsius"/>`);
  parts.push(`    <GenericAttributes Number="2">`);
  parts.push(`      <GenericAttribute Name="ProjectId" Value="${xmlEscape(projectId)}"/>`);
  parts.push(`      <GenericAttribute Name="Revision" Value="${xmlEscape(revision)}"/>`);
  parts.push(`    </GenericAttributes>`);
  parts.push(`  </PlantInformation>`);

  for (const eq of equipment) {
    const attrs = eq.attributes || {};
    const et = (attrs.equipment_type || "generic").toLowerCase();
    const meta = EQUIPMENT_CLASS[et] || EQUIPMENT_CLASS.generic;
    const eid = dexpiId(eq.object_id, "eq");
    parts.push(`  <Equipment`);
    parts.push(`    ID="${eid}"`);
    parts.push(`    TagName="${xmlEscape(eq.tag)}"`);
    parts.push(`    ComponentClass="${meta.cls}"`);
    parts.push(`    ComponentClassURI="${meta.uri}">`);
    const ga = genericAttributes(attrs, eqMap);
    if (ga) parts.push(ga);
    parts.push(`    <Nozzle ID="${nozzleId(eq.object_id, "in")}" TagName="N1" ComponentClass="Nozzle" ComponentClassURI="${RDL}/RDS415214"/>`);
    parts.push(`    <Nozzle ID="${nozzleId(eq.object_id, "out")}" TagName="N2" ComponentClass="Nozzle" ComponentClassURI="${RDL}/RDS415214"/>`);
    parts.push(`  </Equipment>`);
  }

  if (pipes.length) {
    parts.push(`  <PipingNetworkSystem`);
    parts.push(`    ID="pns_main"`);
    parts.push(`    TagName="${xmlEscape(projectId)}-Piping"`);
    parts.push(`    ComponentClass="PipingNetworkSystem"`);
    parts.push(`    ComponentClassURI="${RDL}/RDS270359">`);

    for (const pipe of pipes) {
      const attrs = pipe.attributes || {};
      const segId = dexpiId(pipe.object_id, "seg");
      const fromTag = attrs.from_tag || "";
      const toTag = attrs.to_tag || "";
      const fromEq = equipment.find((e) => e.tag === fromTag);
      const toEq = equipment.find((e) => e.tag === toTag);
      parts.push(`    <PipingNetworkSegment`);
      parts.push(`      ID="${segId}"`);
      parts.push(`      TagName="${xmlEscape(pipe.tag)}"`);
      parts.push(`      ComponentClass="PipingNetworkSegment"`);
      parts.push(`      ComponentClassURI="${RDL}/RDS267704">`);
      const ga = genericAttributes(attrs, pipeMap);
      if (ga) parts.push(ga.replace(/^    /gm, "      "));
      if (fromEq) {
        parts.push(`      <Connection FromID="${nozzleId(fromEq.object_id, "out")}" ToID="${segId}_start"/>`);
      }
      if (toEq) {
        parts.push(`      <Connection FromID="${segId}_end" ToID="${nozzleId(toEq.object_id, "in")}"/>`);
      }
      parts.push(`    </PipingNetworkSegment>`);
    }
    parts.push(`  </PipingNetworkSystem>`);
  }

  for (const inst of instruments) {
    const iid = dexpiId(inst.object_id, "inst");
    parts.push(`  <InstrumentationLoopFunction`);
    parts.push(`    ID="${iid}"`);
    parts.push(`    TagName="${xmlEscape(inst.tag)}"`);
    parts.push(`    ComponentClass="ProcessInstrumentationFunction"`);
    parts.push(`    ComponentClassURI="${RDL}/RDS416842">`);
    const ga = genericAttributes(inst.attributes, instMap);
    if (ga) parts.push(ga);
    parts.push(`  </InstrumentationLoopFunction>`);
  }

  for (const valve of valves) {
    const vt = (valve.attributes?.valve_type || "gate").toLowerCase();
    const vmeta = VALVE_CLASS[vt] || VALVE_CLASS.gate;
    const vid = dexpiId(valve.object_id, "valve");
    parts.push(`  <OperatedValve`);
    parts.push(`    ID="${vid}"`);
    parts.push(`    TagName="${xmlEscape(valve.tag)}"`);
    parts.push(`    ComponentClass="${vmeta.cls}"`);
    parts.push(`    ComponentClassURI="${vmeta.uri}">`);
    const ga = genericAttributes(valve.attributes, valveMap);
    if (ga) parts.push(ga);
    parts.push(`  </OperatedValve>`);
  }

  for (const psv of psvs) {
    const pid = dexpiId(psv.object_id, "psv");
    parts.push(`  <SafetyValveOrFitting`);
    parts.push(`    ID="${pid}"`);
    parts.push(`    TagName="${xmlEscape(psv.tag)}"`);
    parts.push(`    ComponentClass="SafetyValve"`);
    parts.push(`    ComponentClassURI="${RDL}/RDS416842">`);
    const ga = genericAttributes(psv.attributes, psvMap);
    if (ga) parts.push(ga);
    parts.push(`  </SafetyValveOrFitting>`);
  }

  if (discipline === "PFD" && streams.length) {
    for (const stream of streams) {
      const sid = dexpiId(stream.object_id, "str");
      parts.push(`  <ProcessConnection`);
      parts.push(`    ID="${sid}"`);
      parts.push(`    TagName="${xmlEscape(stream.tag)}"`);
      parts.push(`    ComponentClass="ProcessConnection"`);
      parts.push(`    ComponentClassURI="${RDL}/RDS414674">`);
      const ga = genericAttributes(stream.attributes, {});
      if (ga) parts.push(ga);
      parts.push(`  </ProcessConnection>`);
    }
  }

  parts.push(`</PlantModel>`);
  return parts.join("\n");
}

export function dexpiExportSummary(model) {
  const objects = model.objects || [];
  return {
    project_id: model.project_id,
    revision: model.revision,
    equipment: objects.filter((o) => o.class === "Equipment").length,
    pipe_segments: objects.filter((o) => o.class === "PipeRun").length,
    instruments: objects.filter((o) => o.class === "Instrument").length,
    valves: objects.filter((o) => o.class === "Valve").length,
    safety_valves: objects.filter((o) => o.class === "SafetyValve").length,
    streams: objects.filter((o) => o.class === "Stream").length,
    format: "DEXPI-Proteus-subset-1.4",
  };
}
